analytics/lib/data_collectors.py

```diff
@@ -0,0 +1,115 @@
+# being pulled out of the db.
+
+# TODO: need to insert valid_ids that don't show up!!
+
+def process_count(table, valid_ids, id_field, value_function, property, time_interval):
+    # check if there is at least one row with the filter args
+    # note that this will create problems for aborted jobs, since e.g. there will be rows for
+    # for some but not all realms, say
+    if not table.object.filter(property = property,
+                               end_time = time_interval.end,
+                               interval = time_interval.interval).exists():
+        # collect the data from zerver and/or analytics tables
+        rows = value_function(time_interval)
+        # insert all the rows into the appropriate analytics table
+        table.objects.bulk_create([table(property = property,
+                                         end_time = time_interval.end,
+                                         interval = time_interval.interval,
+                                         **row)
+                                   for row in rows if row[id_field] in valid_ids])
+
+
+def process_aggregate_count(table, valid_ids, id_field, aggregate_function, property, time_interval):
+    return process_count(table, valid_ids, id_field,
+                         aggregate_function(property, time_interval),
+                         property, time_interval)
+
+## aggregators. Possibly to be written more generally once we have a StreamCount, and more intervals
+
+def aggregate_user_to_realm(property, time_interval):
+    return UserCount.objects \
+                    .filter(property = property,
+                            end_time = time_interval.end,
+                            interval = time_interval.interval) \
+                    .values('realm') \
+                    .annotate(realm_id=F('realm'), value=Sum('value')) \
+                    .values('realm_id', 'value')
+
+def aggregate_realm_hour_to_day(property, time_interval):
+    return RealmCount.objects \
+                     .filter(interval = 'hour',
+                             property = property,
+                             end_time__gt = time_interval.end - timedelta(days=1),
+                             end_time__lte = time_interval.end) \
+                     .values('realm') \
+                     .annotate(realm_id=F('realm'), value=Sum('value')) \
+                     .values('realm_id', 'value')
+
+def aggregate_user_hour_to_day(property, time_interval):
+    return UserCount.objects \
+                    .filter(interval = 'hour',
+                            property = property,
+                            end_time__gt = time_interval.end - timedelta(days=1),
+                            end_time__lte = time_interval.end) \
+                     .values('user', 'realm') \
+                     .annotate(userprofile_id=F('user'), realm_id=F('realm'), value=Sum('value')) \
+                     .values('userprofile_id', 'realm_id', 'value')
```

Comment:
Just to make sure you know you can do this, with Django's query system, queries are lazily evaluated.  So you could write something like:

`base_query = UserCount.objects.filter(property=property, ...)`
and then later build a more complex query from it using e.g.
`return base_query.filter(... more_Stuff ...)` 
and only one database query would eventually be made.  

It's not immediately clear to me whether that is useful here, but I figured I'd mention it since it is what you'd want to reuse parts of queries between different functions.